# 💥 Avence de proyecto:💥

<a style="background-color: #ff4800; color: #000000; border: none; padding: 10px 15px; border-radius: 5px;">Parte 1</a>
<a style="background-color: #ff4800; color: #000000; border: none; padding: 10px 15px; border-radius: 5px;">28 de Octubre del 2023</a>
<a style="background-color: #ff4800; color: #000000; border: none; padding: 10px 15px; border-radius: 5px;">Cursos PyM</a>
<a style="background-color: #ff4800; color: #000000; border: none; padding: 10px 15px; border-radius: 5px;">Luis Fernando Apáez Álvarez</a>
<a href="#📘-formulario-📘" style="background-color: #fefe04; color: #000000; border: none; padding: 10px 15px; border-radius: 5px;">Formulario</a>
        

---

## 📘 Generación de información 📘

Generaremos una dataframe con la información de venttas de la papelería para una fecha determinada.


In [ ]:
# Lista con las letras del abecedario
abcdario = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
                'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                'U', 'V', 'W', 'X', 'Y', 'Z']

# Lista con las sucursales de la papaleria
papelerias = ['Xochimilco', 'Cuemanco', 'Coapa', 'Milpa Alta', 'CU', 'Zócalo',
              'Narvarte', 'Santa Fé', 'Polanco', 'Centro']

# Lista con los productos que se vederan en la papeleria
lineas = ['Cuadernos', 'Libretas', 'Lápices', 'Plumones', 'Borradores', 'Sacapuntas',
        'Laptops', 'Tablets', 'Mochilas', 'Bolsas', 'Cajas', 'Pegamento', 'Tijeras',
        'Monitores', 'Teclados', 'Mouse', 'Audífonos', 'Cables', 'Cargadores', 'Baterías',
        'Pc', 'Uniformes', 'Pinturas', 'Pinceles', 'Papel', 'Cartulinas']

In [ ]:
# Recordemos que podemos acceder a los elementos de una lista por medio de su índice
# y que el primer elemento tiene índice 0
print(abcdario[0])
print(papelerias[5])
print(lineas[10])

A
Zócalo
Cajas


In [ ]:
# numero total de elementos de una lista cin la funcion len()
len(abcdario)

26

In [ ]:
# como la numeracion en Python termina en 1 menos que el numero de elementos
# el indice del ultimo elemento de una lista es 25
print(abcdario[25])

# o alternativamente
print(abcdario[len(abcdario)-1])

Z
Z


Ahora bien, generaremos 3 número aleatorios entre el 0 y 25 (o ``len(abcdario) - 1``) que serán las posiciones de las letras que vamos a tomar. Con lo anterior podremos obtener 3 letras de manera aleatoria, ejemplo:


In [ ]:
# importacion de la libreria randomq ue trae funcionalidades para generar numeros aleatorios
import random as r

# con la funcion randint(a,b) de la libreria random podemos generar un numero aleatorio
# entre a y b. De tal manera, generamos 3 numeros aleatorios entre 0 y 25 (o len(abcdario)-1)
r1 = r.randint(0, len(abcdario)-1)
r2 = r.randint(0, len(abcdario)-1)
r3 = r.randint(0, len(abcdario)-1)

# Accedemos a los elementos de la lista abcdario con los indices generados aleatoriamente
print(abcdario[r1])
print(abcdario[r2])
print(abcdario[r3])

L
M
E


In [ ]:
# Recordemos que podemos juntar cadenas de texto con el operador +
print(abcdario[r1] + abcdario[r2] + abcdario[r3])

LME


El código anterior lo podemos dejar mejor estructura de la siguiente manera:

In [ ]:
# en vez de calcular 3 veces len(abcdario) podemos guardar el resultado en una variable
# para que se calcule solo una vez
longitud_abc = len(abcdario) - 1

r1 = r.randint(0, longitud_abc)
r2 = r.randint(0, longitud_abc)
r3 = r.randint(0, longitud_abc)

clave = abcdario[r1] + abcdario[r2] + abcdario[r3]
clave

'NRY'

Notamos que los códigos generados son diferentes, lo cual en efecto debe pasar dado que estamos generando números aleatorios cada vez que ejecutamos el código. SI queremos generar un número aleatorio que después de volver a ejecutar no cambie, deberemos de fijar una _semilla aleatoria_:

In [ ]:
# dentro de la funcion seed podemos poner cualquier numero entero
r.seed(0)

# mejoraremos aun mas el codigo
longitud_abc = len(abcdario) - 1

clave = ''
for i in range(3):
    clave += abcdario[r1]

clave

'NNN'

¿Se puede mejorar más? 👀

In [ ]:
longitud_abc = len(abcdario) - 1
clave = ''.join([abcdario[r.randint(0, longitud_abc)] for i in range(3)])
clave

'MYN'

In [ ]:
# finalmente al codigo MYN le agregaremos un numero aleatorio de 3 digitos
longitud_abc = len(abcdario) - 1

r1 = r.randint(0, longitud_abc)
r2 = r.randint(0, longitud_abc)
r3 = r.randint(0, longitud_abc)
num1 = r.randint(0, 10)
num2 = r.randint(0, 10)
num3 = r.randint(0, 10)

# la funcion str() convierte un numero a cadena de texto, recordemos que solo podemos
# juntar cadenas de texto con el operador + y no podemos combinar entre cadenas de texto
# y numeros
clave = abcdario[r1] + abcdario[r2] + abcdario[r3] + str(num1) + str(num2) + str(num3)
clave

'BIQ764'

Pasaremos a generar un precio aleatorio para el producto con la clave anterior

In [ ]:
# para generar un numero decimal aleatorio haremos lo siguiente:
# Generamos un numero aleatorio entre el 0 y el 1, luego lo multiplicamos por un numero
# aleatorio entero entre 1 y 50000
precio = r.random() * r.randint(1, 50000)
precio

22710.394680796744

In [ ]:
# podemos redondear a 3 decimales
precio = round(precio, 3)
precio

22710.395

Una vez el precio y la clave generada, escogeremos un producto de nuestra lista ``lineas`` de manera aleatoria:

In [ ]:
# para escoger aleatoriamente un elemento de una lista usamos la funcion choice()
producto = r.choice(lineas)
producto

'Cargadores'

Pongamos ahora una fecha:

In [ ]:
dia = '2020-01-01'
dia

'2020-01-01'

Pero nosotros lo queremos en el formato ``ddmmaaaa``, para lo cual jugaremos con los índices de la cadena de texto:

In [ ]:
print(dia[-2:])
print(dia[5:7])
print(dia[:4])

01
01
2020


In [ ]:
# todo junto
dia_str = dia[-2:] + dia[5:7] + dia[:4]
dia_str

'01012020'

# 📘 Diccionarios 📘

Queremos guardar la información anterior de la siguiente manera:

* Clave-Precio: Para cada clave de producto queremos tener su precio asignado y guardado.
* Clave-Producto: Para cada clave de producto queremos tener su nombre asignado y guardado.

La estructura de _Diccionario_ en Python nos permite guardar información así como lo necesitamos pues nos permite guardar un valor asociado a una clave. La sintaxis es la siguiente:


In [ ]:
# lo siguente es un comentario multilinea
"""
diccionario = {clave1: valor1,
               clave2: valor2,
               clave3: valor3}
"""

'\ndiccionario = {clave1: valor1, \n               clave2: valor2, \n               clave3: valor3}\n'

De tal manera, en un diccionario pondremos como claves, justamente, las clave de los productos y como valores, el precio de dicho producto. De la misma manera crearemos un diccionario para guardar el nombre de los productos asociado a su clave. Antes de hacerlo veamos un ejemplo de un diccionario:

In [ ]:
# diccionario de numeros
nums = {'uno': 1,
        'dos': 2,
        'tres': 3}

# donde las claves del diccionario son las palabras y los valores son los numeros
# asi como en una lista, podemos acceder a los valores de un diccionario por medio de sus claves

print(nums['uno'])
print(nums['dos'])
print(nums['tres'])

1
2
3


In [ ]:
# otro ejemplo
lenguajes = {'Python': 'Lenguaje de programación de alto nivel',
             'Java': 'Lenguaje de programación de alto nivel',
              'C': 'Lenguaje de programación de medio nivel',
              'R': 'Lenguaje de programación de alto nivel'}

print(lenguajes['Python'])
print(lenguajes['C'])

Lenguaje de programación de alto nivel
Lenguaje de programación de medio nivel


Podemos agregar nuevos elementos a un diccionario de la siguiente manera:


In [ ]:
# vemos el diccionario completo
print(nums)

{'uno': 1, 'dos': 2, 'tres': 3}


In [ ]:
# no tiene la clave, por ejemplo, 'cuatro', de modo que si escribirmos
nums['cuatro'] = 4

# estaremos agregando la clave 'cuatro' con el valor 4
nums

{'uno': 1, 'dos': 2, 'tres': 3, 'cuatro': 4}

# 📘 Generación de los datos 📘

In [ ]:
%%time
# Generaremos 1000 claves aleatorias para el diccionario
# Podemos crear diccionarios vacios que luego iremos llenando
clave_precio = {}
clave_producto = {}

# iteramos 1000 veces
for i in range(1000):
    #####################################################################################
    ######################### GENERAR CLAVE #############################################
    r1 = r.randint(0, longitud_abc)
    r2 = r.randint(0, longitud_abc)
    r3 = r.randint(0, longitud_abc)

    num1 = r.randint(0, 10)
    num2 = r.randint(0, 10)
    num3 = r.randint(0, 10)

    clave = abcdario[r1] + abcdario[r2] + abcdario[r3] + str(num1) + str(num2) + str(num3)
    #####################################################################################
    ######################### GENERAR PRECIO ############################################
    precio = r.random() * r.randint(1, 50000)
    precio = round(precio, 3)
    #####################################################################################
    ######################### ESCOGER UN PRODUCTO #######################################
    producto = r.choice(lineas)
    #####################################################################################
    ######################### LLENAR INFO EN LOS DICCIONARIO ############################
    clave_precio[clave] = precio
    clave_producto[clave] = producto

CPU times: total: 15.6 ms
Wall time: 11 ms


En vez de ver la información como un diccionario, la veremos como un dataframe de pandas 🐼

In [ ]:
# antes de ello veamos que hace la funcion items() de un diccionario
clave_precio.items()

dict_items([('GQE421', 10152.382), ('WZT241', 32715.438), ('PRD565', 8186.496), ('POQ408', 5599.474), ('MWZ10100', 13359.549), ('XKW139', 2070.493), ('ROC158', 32000.958), ('JRJ185', 28839.457), ('ZTR947', 2311.42), ('SHJ232', 1418.968), ('PCC1022', 4856.984), ('RVM848', 11447.211), ('SNS477', 30298.66), ('LCK917', 12899.792), ('HAX413', 4157.029), ('NBD230', 33962.788), ('TVC0110', 7115.899), ('MCL109', 262.299), ('DPG0100', 22135.143), ('ICH1104', 4139.774), ('QOB916', 4684.662), ('PSF1030', 8177.494), ('KQI197', 576.415), ('VNS8410', 15392.523), ('ERW071', 1006.07), ('IEH759', 6776.934), ('UTE466', 4385.725), ('TGW523', 6552.16), ('WVS606', 32418.084), ('YVW027', 2937.173), ('OQP890', 28649.045), ('JOB638', 41466.95), ('XEA6106', 4424.583), ('WYA1089', 762.033), ('UGJ421', 12364.791), ('CAI714', 4554.444), ('ULD240', 570.341), ('IRK590', 38901.241), ('UPW10710', 15351.024), ('GMS402', 3300.007), ('ZLW159', 630.209), ('ESJ568', 976.196), ('XHB428', 14453.715), ('KJN118', 28227.447), 

cada tupla será una fila del dataframe

In [ ]:
# importacion necesaria para crear un dataframe
import pandas as pd

# con la funcion DataFrame() de la libreria pandas podemos crear un dataframe
df1 = pd.DataFrame(clave_precio.items(), columns=['clave', 'precio'])
df1.head()

clave     precio
0    GQE421  10152.382
1    WZT241  32715.438
2    PRD565   8186.496
3    POQ408   5599.474
4  MWZ10100  13359.549

In [ ]:
# Creamos otro diccionario
df2 = pd.DataFrame(clave_producto.items(), columns=['clave', 'producto'])
df2.head()

clave  producto
0    GQE421    Cables
1    WZT241     Cajas
2    PRD565    Cables
3    POQ408  Pinceles
4  MWZ10100   Tablets

Nos gustaría tener un dataframe con la siguiente estructura:

| clave | producto | precio |
|-------|----------|--------|
|GQE421 | Cables   |10152.382|

en vez de tener dos dataframes, para lo cual podemos juntar los dataframes df1 y df2 tomando como criterio que la clave debe ser la misma

In [ ]:
# union de los dos dataframes tomando como criterio que la clave sea la misma
df_final = df1.merge(df2, on='clave')
df_final.head()

clave     precio  producto
0    GQE421  10152.382    Cables
1    WZT241  32715.438     Cajas
2    PRD565   8186.496    Cables
3    POQ408   5599.474  Pinceles
4  MWZ10100  13359.549   Tablets

In [ ]:
# Reordenamos las columnas a como deseamos que queden
df_final = df_final[['clave', 'producto', 'precio']]

# vemos los ultimos 5 elementos del dataframe
df_final.tail()

clave producto     precio
995   DKH578  Tablets   4941.076
996   MCP707  Tablets   2546.689
997   JTQ217  Tijeras  19472.595
998   FAB117  Laptops   2823.760
999  MKI0910    Mouse  22278.875

In [ ]:
# veamos informacion sobre el dataframe
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   clave     1000 non-null   object 
 1   producto  1000 non-null   object 
 2   precio    1000 non-null   float64
dtypes: float64(1), object(2)
memory usage: 23.6+ KB


In [ ]:
df_final.describe()

precio
count   1000.000000
mean   12312.942786
std    10922.429325
min       16.895000
25%     3328.633750
50%     8907.735500
75%    18439.833000
max    49128.989000

Agregremos dos columnas: una con alguna sucursal y otra con la fecha

In [ ]:
print(dia)

2020-01-01


In [ ]:
df_final['dia'] = dia
df_final.head()

clave  producto     precio         dia
0    GQE421    Cables  10152.382  2020-01-01
1    WZT241     Cajas  32715.438  2020-01-01
2    PRD565    Cables   8186.496  2020-01-01
3    POQ408  Pinceles   5599.474  2020-01-01
4  MWZ10100   Tablets  13359.549  2020-01-01

In [ ]:
# escogemos de manera aletoria una sucursal
sucursal = r.choice(papelerias)
sucursal

'Polanco'

In [ ]:
# hacemos lo mismo que antes
df_final['sucuarsal'] = sucursal
df_final.head()

clave  producto     precio         dia sucuarsal
0    GQE421    Cables  10152.382  2020-01-01   Polanco
1    WZT241     Cajas  32715.438  2020-01-01   Polanco
2    PRD565    Cables   8186.496  2020-01-01   Polanco
3    POQ408  Pinceles   5599.474  2020-01-01   Polanco
4  MWZ10100   Tablets  13359.549  2020-01-01   Polanco

# 📘 Exportación a diferentes fuentes 📘

In [ ]:
# exportamos a un excel
# df_final.to_excel('ventas.xlsx', index=False)

# exportamos a un csv
# df_final.to_csv('ventas.csv', index=False)

# exportamos como json
# df_final.to_json('ventas.json', orient='records')

In [ ]:
# habra repetidos?
# numero total de filas del dataframe
print(len(df_final))

# quitamos duplicados
print(len(df_final.drop_duplicates()))

1000
1000


In [ ]:
# df_final.to_html('ventas.html', index=False)